### PR TITLE
cleanup(cri): enhance robustness for container image retrieval

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Attempt to enhance robustness for container image retrieval.

Did some production analyses and it seems the issue of `container.image.repository` occasionally being just `sha256` is extremely rare. I think that @gnosek previous patch https://github.com/falcosecurity/libs/pull/771 helped a lot, but most end users could not yet test it as it hasn't yet been part of a Falco release. In contrast I have that patch already running.

Since this is very critical for detection and incident response tried harder to add even more robustness. As you see in the minor code changes there are still some more places from where we can try to retrieve the container image. Fingers crossed all patches together eliminate remaining corner cases.

Verified the schemas locally with both `containerd` and `crio` runtimes, plus queried the schema for `containerd` using `crictl` on a real Kubernetes test server as well.

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/falco/issues/1187

CC @gnosek @leogr @sigurdfalk

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
cleanup(cri): enhance robustness for container image retrieval
```

### CODE REVIEW

File: userspace/libsinsp/cri.cpp [hunk 1]
```diff
@@ -143,6 +143,33 @@ bool cri_interface::parse_cri_image(const runtime::v1alpha2::ContainerStatus &st
...
 	}
 
+	if(image_name.empty() || strncmp(image_name.c_str(), "sha256", 6) == 0)
+	{
+		/* Retrieve image_name from annotations as backup when image name may start with sha256,
+		brute force try each schema we know of for containerd and crio container runtimes. */
+
+		Json::Value root;
+		Json::Reader reader;
+		if(reader.parse(info.find("info")->second, root))
```
Comment (gnosek) [inline userspace/libsinsp/cri.cpp:153]: One question: what happens if "info" is not found in `info` (uhh... naming? :D)? Just for testing, can you change "info" to "whatever" to simulate an unsupported runtime and see if it crashes?

File: userspace/libsinsp/cri.cpp [hunk 2]
```diff
@@ -143,6 +143,33 @@ bool cri_interface::parse_cri_image(const runtime::v1alpha2::ContainerStatus &st
...
+			Json::Value jvalue;
+			jvalue = root["runtimeSpec"]["annotations"]["io.kubernetes.cri.image-name"];
+			if(jvalue.isNull())
+			{
+				jvalue = root["runtimeSpec"]["annotations"]["io.kubernetes.cri-o.Image"];
+			}
+			if(jvalue.isNull())
+			{
+				jvalue = root["runtimeSpec"]["annotations"]["io.kubernetes.cri-o.ImageName"];
+			}
```
Comment (gnosek) [inline userspace/libsinsp/cri.cpp:164]: Same, what happens if there's no runtimeSpec or no annotations? Again, temporarily introducing typos could help :)

**Review by gnosek — COMMENTED** [review]
I'm not personally a fan of accessing k8s data when trying to get some basic container info but I understand I've already lost this battle, so LGTM, modulo the comments

File: userspace/libsinsp/cri.cpp (hunk 1 above)
Comment (incertum) [reply userspace/libsinsp/cri.cpp]: Added, the other case you pointed out are fine. Also missed that when I added `container.cni.json`, therefore aded checks for both cases. Thanks @gnosek!

File: userspace/libsinsp/cri.cpp (hunk 2 above)
Comment (FedeDP) [reply userspace/libsinsp/cri.cpp]: JsonCPP returns null (just tested); i think the code is fine.

**Review by FedeDP — APPROVED** [review]
/approve
LGTM!

**Review by leogr — APPROVED** [review]
LGTM, now!

Thank you!

**Comment by leogr** [comment]
Closing and reopening the PR to trigger the CI (the issue was likely due to the GH outage).
/close
